src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class DefaultArrayPool<T> : ArrayPool<T>
+    {
+        private const int MinimiumArraySize = 16;
+        private int _maxArrayLength;
+        private int _numberOfArraysPerBucket;
+        private bool _isTObjectType;    // cache this to only do the reflection call once
+        private ArrayPoolBucket<T>[] _buckets;
+        private int _rentedBufferCount;
+        private int _totalBufferCount;
+
+        public override int RentedBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _rentedBufferCount);
+            }
+        }
+
+        public override int TotalBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _totalBufferCount);
+            }
+        }
+
+        internal DefaultArrayPool(int maxLength, int arraysPerBucket)
+        {
+            // Our bucketing algorithm has a minimum length of 16
+            if (maxLength < MinimiumArraySize)
+                throw new ArgumentOutOfRangeException(nameof(maxLength), SR.Maximum_Array_Size_Under_Minimum_Supported_Value);
```

Comment:
You're right, we probably shouldn't throw here since that's an implementation detail and should not be part of the contract. I'll fix this.

Removed all nameof calls